### ISSUE TITLE
Upgrade response

### ISSUE THREAD

**Issue by fafhrd91** [MagicStack/httptools#3]
here test code. also i am not sure how to get data after end of message back. i am about `\x89\x04data`

```python

import httptools

import unittest
from unittest import mock

UPGRADE_RESPONSE = b'''HTTP/1.1 101 Switching Protocols
UPGRADE: websocket
SEC-WEBSOCKET-ACCEPT: rVg+XakFNFOxk3ZH0lzrZBmg0aU=
TRANSFER-ENCODING: chunked
CONNECTION: upgrade
DATE: Sat, 07 May 2016 23:44:32 GMT
SERVER: Python/3.4 aiohttp/1.0.3
\r\n\x89\x04data'''


class TestResponseParser(unittest.TestCase):

    def test_parser_upgrade_response(self):
        m = mock.Mock()

        headers = {}
        m.on_header.side_effect = headers.__setitem__

        p = httptools.HttpResponseParser(m)
        try:
            p.feed_data(UPGRADE_RESPONSE)
        except:
            pass

        self.assertEqual(p.get_http_version(), '1.1')
        self.assertEqual(p.get_status_code(), 101)

        m.on_status.assert_called_once_with(b'Switching Protocols')

        m.on_headers_complete.assert_called_once_with()
        self.assertEqual(m.on_header.call_count, 6)
        self.assertEqual(len(headers), 6)

        m.on_message_complete.assert_called_once_with()

```

**Comment by 1st1** [comment]
On how to get the data, see the existing unittest for upgrade. Although I think I'll remove the exception in favour of a callback, as you suggested.

**Comment by fafhrd91** [comment]
i think there is bug in offset calculation

```python

import httptools

import unittest
from unittest import mock

UPGRADE_RESPONSE = b'''HTTP/1.1 101 Switching Protocols\r\nUPGRADE: websocket\r\nSEC-WEBSOCKET-ACCEPT: rVg+XakFNFOxk3ZH0lzrZBmg0aU=\r\nTRANSFER-ENCODING: chunked\r\nCONNECTION: upgrade\r\nDATE: Sat, 07 May 2016 23:44:32 GMT\r\nSERVER: Python/3.4 aiohttp/1.0.3\r\n\r\n\x89\x04data'''


class TestResponseParser(unittest.TestCase):

    def test_parser_upgrade_response(self):
        m = mock.Mock()

        headers = {}
        m.on_header.side_effect = headers.__setitem__

        p = httptools.HttpResponseParser(m)
        try:
            p.feed_data(UPGRADE_RESPONSE)
        except httptools.HttpParserUpgrade as ex:
            offset = ex.args[0]
        else:
            self.fail('HttpParserUpgrade was not raised')

        self.assertEqual(UPGRADE_RESPONSE[offset:], b'\x89\x04data')

        self.assertEqual(p.get_http_version(), '1.1')
        self.assertEqual(p.get_status_code(), 101)

        m.on_status.assert_called_once_with(b'Switching Protocols')

        m.on_headers_complete.assert_called_once_with()
        self.assertEqual(m.on_header.call_count, 6)
        self.assertEqual(len(headers), 6)

        m.on_message_complete.assert_called_once_with()
```
got this ```AssertionError: b'\x04data' != b'\x89\x04data'```

**Comment by 1st1** [comment]
Transfer encoding is set to `chunked`. If I remove the `transfer-encoding` header everything works as expected.

**Comment by 1st1** [comment]
Are you sure that this response is valid?

**Comment by fafhrd91** [comment]
i am not sure, but `autobahn` test suit passes for default websocket implementation.

**Comment by 1st1** [comment]
Is this response something that autobahn tests for?

To me this seems like an invalid response -- you essentially ask the protocol to upgrade, at the same time specifying the transfer-encoding.  If this is a real-world situation then we probably need to fix http-parser.c

**Comment by 1st1** [comment]
Alright, seems that http-parser support this -- have to fix the httptools. Working on it.

**Comment by 1st1** [comment]
Should be fixed in `v0.0.9`

**Comment by fafhrd91** [comment]
thanks, works!
